### PR TITLE
porcelain.checkout continuation

### PR DESCRIPTION
**Pull request by hsorby**
Carrying on from work started #898.

### CODE REVIEW

File: dulwich/errors.py [hunk 1]
```diff
@@ -91,6 +91,10 @@ class NotBlobError(WrongObjectException):
     type_name = "blob"
 
 
+class DirNotCleanError(Exception):
```
Comment (jelmer) [inline dulwich/errors.py:94]: Per the comment in this file, please don't add more errors here but rather near the code that raises them.

File: dulwich/porcelain.py [hunk 2]
```diff
@@ -303,10 +302,10 @@ def check_diverged(repo, current_sha, new_sha):
 
 
 def archive(
-    repo,
-    committish=None,
-    outstream=default_bytes_out_stream,
-    errstream=default_bytes_err_stream,
+        repo,
```
Comment (jelmer) [inline dulwich/porcelain.py:305]: Please don't include formatting changes in the same PR, it makes it harder to review.

File: dulwich/tests/test_porcelain.py [hunk 3]
```diff
@@ -1577,10 +1577,230 @@ def test_resetfile_with_dir(self):
...
+        porcelain.checkout(target_repo, b"master")
+        self.assertEqual(b"master", porcelain.active_branch(target_repo))
+        self.assertFalse(os.path.isfile(os.path.join(target_repo.path, 'bar')))
+
+        porcelain.checkout(target_repo, b"origin/foo")
+        self.assertEqual(b"foo", porcelain.active_branch(target_repo))
+        self.assertTrue(os.path.isfile(os.path.join(target_repo.path, 'bar')))
+
+
+def tree(root_dir):
```
Comment (jelmer) [inline dulwich/tests/test_porcelain.py:1795]: This isn't actually used?

File: dulwich/porcelain.py [hunk 4]
```diff
@@ -1859,6 +1858,95 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
...
+            target_entry = None
+
+        if current_entry or target_entry:
+            tracked_changes.append(file.decode())
+
+    return tracked_changes
+
+
+def checkout(repo, target: bytes, force: bool = False):
+    """switch branches or restore working tree files
```
Comment (jelmer) [inline dulwich/porcelain.py:1889]: This should probably be separate functions for the different modes of using checkout; i.e. one for checking out a new branch, and one for refreshing a file in the checkout.

**Review by jelmer — CHANGES_REQUESTED** [review]

File: dulwich/porcelain.py [hunk 5]
```diff
@@ -1859,6 +1858,95 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
...
+    Args:
+      repo: dulwich Repo object
+      target: branch name or commit sha to checkout
+      force: true or not to force checkout
+    """
+    # Check repo status.
+    if not force:
+        index = repo.open_index()
+        for file in get_tree_changes(repo)['modify']:
+            if file in index:
```
Comment (jelmer) [inline dulwich/porcelain.py:1899]: this check seems odd; when would a file have changed but not be in the index?

File: dulwich/porcelain.py [hunk 6]
```diff
@@ -1859,6 +1858,95 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
...
+        else:
+            update_head(repo, target, detached=True)
+
+    # Un-stage files in the current_tree or target_tree.
+    tracked_changes = _determine_tracked_changes(repo, current_tree, target_tree)
+    repo.unstage(tracked_changes)
+
+    repo.reset_index(target_tree.id)
+
+    # Remove the untracked file which are in the current_file_set.
```
Comment (jelmer) [inline dulwich/porcelain.py:1940]: Wouldn't it be better to get the list of files in the previous index and diff that against the new index?

File: dulwich/porcelain.py (hunk 5 above)
Comment (hsorby) [reply dulwich/porcelain.py:1899]: Yes, I have come to the same conclusion but wanted to hear someone else's opinion who knows more about this than I.  I will remove this check.

File: dulwich/porcelain.py (hunk 6 above)
Comment (hsorby) [reply dulwich/porcelain.py:1940]: Possibly, I am not super familiar with this.  Any hints or existing code that does this that I can look at?

File: dulwich/tests/test_porcelain.py (hunk 3 above)
Comment (hsorby) [reply dulwich/tests/test_porcelain.py:1795]: Whoops, used it for debugging.

File: dulwich/porcelain.py (hunk 2 above)
Comment (hsorby) [reply dulwich/porcelain.py:305]: Sorry, PyCharm auto-formatted.  Do you want me to remove these changes?

File: dulwich/porcelain.py (hunk 4 above)
Comment (jelmer) [inline dulwich/porcelain.py:1889]: Please document here that this is a naive implementation of checkout:

* it doesn't allow any unknowns or changed files to exist in the tree
* it's fairly inefficient (since it scans the entire tree for unknown files)

File: dulwich/errors.py (hunk 1 above)
Comment (hsorby) [reply dulwich/errors.py:94]: Okay.

File: dulwich/porcelain.py (hunk 6 above)
Comment (jelmer) [reply dulwich/porcelain.py:1940]: calling repo.open_index() will get you the current index (before making changes to it with repo.unse or repo.reset_index, obviously)

the index object provides an iterator that returns all the relative paths in the index, as bytestrings.

File: dulwich/porcelain.py (hunk 4 above)
Comment (jelmer) [reply dulwich/porcelain.py:1889]: Since you're only implementing the branch version here, perhaps call it "checkout_branch" rather than "checkout" ?

File: dulwich/errors.py (hunk 1 above)
Comment (hsorby) [reply dulwich/errors.py]: [12118cc](https://github.com/jelmer/dulwich/pull/1148/commits/12118cc43385f143853c4a2306d23dc08f994aff)

File: dulwich/porcelain.py (hunk 2 above)
Comment (hsorby) [reply dulwich/porcelain.py]: Done.

File: dulwich/tests/test_porcelain.py (hunk 3 above)
Comment (hsorby) [reply dulwich/tests/test_porcelain.py]: [90180da](https://github.com/jelmer/dulwich/pull/1148/commits/90180dadacaf2fff71cb9373dd9e4614a9a1eb4f)

**Comment by hsorby** [comment]
Any idea on what this error is about, coming from the Windows testing:
```
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 
'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpff5pb4ee\\.git\\objects\\pack\\pack-d523f7421691f4381d40ab668eabc12a477c17a5.idx'
```

**Review by jelmer — CHANGES_REQUESTED** [review]
Please revert the unnecessary formatting so that the PR just contains the minimal changes necessary.

File: dulwich/porcelain.py [hunk 7]
```diff
@@ -1859,6 +1863,114 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
...
+        repo_index = repo.open_index()
+        for entry in iter_tree_contents(repo.object_store, target_tree.id):
+            dealt_with.append(entry.path)
+            if entry.path in changes:
+                continue
+            full_path = os.path.join(os.fsencode(repo.path), entry.path)
+            blob = repo.object_store[entry.sha]
+            ensure_dir_exists(os.path.dirname(full_path))
+            st = build_file_from_blob(blob, entry.mode, full_path)
+            st_tuple = (
```
Comment (jelmer) [inline dulwich/porcelain.py:1936]: There should be no need to construct st_tuple here, you can just pass "st" to index_entry_from_stat. if necessary, you can pass a custom mode to index_entry_from_stat.

File: dulwich/porcelain.py [hunk 8]
```diff
@@ -1859,6 +1863,114 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
...
+                pass
+            update_head(repo, LOCAL_BRANCH_PREFIX + checkout_target)
+        else:
+            update_head(repo, target, detached=True)
+
+    return checkout_target
+
+
+def checkout_branch(repo, target: Union[bytes, str], force: bool = False):
+    """switch branches or restore working tree files
```
Comment (jelmer) [inline dulwich/porcelain.py:1891]: please add a comment here about the performance here being subpar, since it first does a diff and only then applies changes

File: dulwich/porcelain.py [hunk 9]
```diff
@@ -1859,6 +1863,114 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
...
+    current_tree = parse_tree(repo, repo.head())
+    target_tree = parse_tree(repo, target)
+
+    if force:
+        repo.reset_index(target_tree.id)
+        _update_head_during_checkout_branch(repo, target)
+    else:
+        status_report = status(repo)
+        changes = list(set(status_report[0]['add'] + status_report[0]['delete'] + status_report[0]['modify'] + status_report[1]))
+        index = 0
```
Comment (jelmer) [inline dulwich/porcelain.py:1908]: couldn't this just be "for change in changes" ?

File: dulwich/porcelain.py (hunk 8 above)
Comment (hsorby) [reply dulwich/porcelain.py]: [a8d9031](https://github.com/jelmer/dulwich/pull/1148/commits/a8d9031ace61135fd343b163843cb2f2280212f8)

File: dulwich/porcelain.py [hunk 10]
```diff
@@ -124,7 +125,6 @@ def write(self, b):
 default_bytes_out_stream = getattr(sys.stdout, "buffer", None) or NoneStream()
 default_bytes_err_stream = getattr(sys.stderr, "buffer", None) or NoneStream()
 
-
```
[the remark this answers is not in the document]
Comment (hsorby) [reply dulwich/porcelain.py]: [a8d9031](https://github.com/jelmer/dulwich/pull/1148/commits/a8d9031ace61135fd343b163843cb2f2280212f8)

File: dulwich/porcelain.py (hunk 7 above)
Comment (hsorby) [reply dulwich/porcelain.py]: [1d5b8d4](https://github.com/jelmer/dulwich/pull/1148/commits/1d5b8d42cb8b679561daf5aae71b08e277d426cf)

File: dulwich/porcelain.py [hunk 11]
```diff
@@ -143,6 +144,10 @@ class TimezoneFormatError(Error):
     """Raised when the timezone cannot be determined from a given string."""
 
 
+class AbortCheckout(Error):
```
Comment (jelmer) [inline dulwich/porcelain.py:147]: CheckoutError perhaps, for consistency with other exception names?

File: dulwich/porcelain.py [hunk 12]
```diff
@@ -1598,7 +1603,7 @@ def active_branch(repo):
         active_ref = r.refs.follow(b"HEAD")[0][1]
         if not active_ref.startswith(LOCAL_BRANCH_PREFIX):
             raise ValueError(active_ref)
-        return active_ref[len(LOCAL_BRANCH_PREFIX) :]
```
Comment (jelmer) [inline dulwich/porcelain.py:1601]: unnecessary whitespace change

File: dulwich/porcelain.py [hunk 13]
```diff
@@ -1859,6 +1864,107 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
...
+                    raise AbortCheckout('Your local changes to the following files would be overwritten by checkout: ' + change.decode())
+            except KeyError:
+                changes.pop(index)
+
+        # Update head.
+        checkout_target = _update_head_during_checkout_branch(repo, target)
+        if checkout_target is not None:
+            target_tree = parse_tree(repo, checkout_target)
+
+        dealt_with = []
```
Comment (jelmer) [inline dulwich/porcelain.py:1933]: could this be a set, for better performance?

**Review by jelmer — CHANGES_REQUESTED** [review]
mostly good to go; two more really minor comments. Please take a look at the failing tests

File: dulwich/porcelain.py (hunk 12 above)
Comment (hsorby) [reply dulwich/porcelain.py:1601]: This one was to fix `PEP 8: E203 whitespace before ':'` PEP8 warning.  Just checking that you want the space before the ':'.

**Comment by hsorby** [comment]
> mostly good to go; two more really minor comments. Please take a look at the failing tests

I don't know why the Windows tests are failing, I need some help on [that](https://github.com/jelmer/dulwich/pull/1148#issuecomment-1451189631).

File: dulwich/porcelain.py (hunk 12 above)
Comment (jelmer) [reply dulwich/porcelain.py:1601]: Not particularly (dulwich is meant to follow PEP8), but it's unrelated to the other changes.

**Comment by jelmer** [comment]
> Any idea on what this error is about, coming from the Windows testing:
> 
> ```
> PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 
> 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmpff5pb4ee\\.git\\objects\\pack\\pack-d523f7421691f4381d40ab668eabc12a477c17a5.idx'
> ```

It probably means that you're not closing the repository (by calling .close()), which means the files stay open and can't be deleted.

File: dulwich/porcelain.py (hunk 11 above)
Comment (hsorby) [reply dulwich/porcelain.py]: [7fca593](https://github.com/jelmer/dulwich/pull/1148/commits/7fca59313dec9a3b2aa64003aeb85e7626726d58)

File: dulwich/porcelain.py (hunk 12 above)
Comment (hsorby) [reply dulwich/porcelain.py]: [daaf927](https://github.com/jelmer/dulwich/pull/1148/commits/daaf92711c725e989d670761a961ffd2ab3d59e5)

File: dulwich/porcelain.py (hunk 13 above)
Comment (hsorby) [reply dulwich/porcelain.py]: [daaf927](https://github.com/jelmer/dulwich/pull/1148/commits/daaf92711c725e989d670761a961ffd2ab3d59e5)

**Comment by hsorby** [comment]
All comments should have been addressed now, all the tests are passing now with latest changes.

**Comment by jelmer** [comment]
Looks good at first glance! I'll take a closer look tomorrow & merge.

**Review by jelmer — APPROVED** [review]
